### ISSUE TITLE
`foreach` derivatives saves redundant tensors

### ISSUE THREAD

**Issue by crcrpar** [pytorch/pytorch#95431]
## Issue description

#89591 and #93901 added the backward of foreach functions but it turned out that the `grad_fn` seems to save unwanted `SaveTensor`s to each Tensor of outputs.

That being said, I'm not quite sure if it'd be an expected behavior where each result Tensor's grad_fn has saved Tensor(s) (and Scalar(s)) with the same index.

## Code example
The following snippet would tell `outputs[0].grad_fn` has saved 2 Tensors even though each output Tensor needs one Tensor to compute its gradient (a.k.a. grad_in).

```python
import torch


def main():
    x, y = [[torch.randn(1, requires_grad=True) for _ in range(2)] for _ in range(2)]
    outputs = torch._foreach_pow(x, y)
    print(f"{len(outputs[0].grad_fn._saved_self) = }")


if __name__ == "__main__":
    main()
```

## System Info
```
Collecting environment information...
PyTorch version: 2.0.0a0+git873bf7d
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.4 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: 13.0.1-++20220120110924+75e33f71c2da-1~exp1~20220120231001.58
CMake version: version 3.25.0
Libc version: glibc-2.31

Python version: 3.10.8 (main, Nov  4 2022, 13:48:29) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.14.0-1042-oem-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 11.7.64
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3070 Ti
Nvidia driver version: 515.43.04
cuDNN version: Probably one of the following:
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn.so.8.5.0
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.5.0
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.5.0
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.5.0
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.5.0
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.5.0
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.5.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False

```

cc @ezyang @gchanan @zou3519 @albanD @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @vincentqb @jbschlosser @janeyx99 @mcarilli @ngimel

**Comment by crcrpar** [comment]
The following is what is currently generated for `_foreach_acos`:
I wrote the logic while keeping changes minimal 

`ForeachAcosBackward0`'s definition:
```cpp
// torch/csrc/autograd/generated/Functions.h
struct TORCH_API ForeachAcosBackward0 : public TraceableFunction {
  using TraceableFunction::TraceableFunction;
  variable_list apply(variable_list&& grads) override;
  std::string name() const override { return "ForeachAcosBackward0"; }
  void release_variables() override {
    std::lock_guard<std::mutex> lock(mutex_);
    self_.clear();
    self_released_ = true;
  }

  std::vector<SavedVariable> self_;
  bool self_released_ = false;
  size_t self_size_;
};

// torch/csrc/autograd/generated/Functions.cpp
variable_list ForeachAcosBackward0::apply(variable_list&& grads) {
  std::lock_guard<std::mutex> lock(mutex_);
  TORCH_CHECK(!self_released_, ERR_BACKWARD_TWICE);
  IndexRangeGenerator gen;
  auto self_ix = gen.range(self_size_);
  variable_list grad_inputs(gen.size());
  auto self = unpack_list(self_);
  if (task_should_compute_output({ self_ix })) {
    std::vector<Tensor> grad_result;
    grad_result.reserve(grads.size());
    for (const auto & i : c10::irange(grads.size())) {
      grad_result.emplace_back(grads[i] * -((-self[i] * self[i] + 1).rsqrt()).conj());
    }
    copy_range(grad_inputs, self_ix, grad_result);
  }
  return grad_inputs;
}
```

Callsite in `VariableType`:
```cpp
// torch/csrc/autograd/generated/VariableTypeEverything.cpp
::std::vector<at::Tensor> _foreach_acos(c10::DispatchKeySet ks, at::TensorList self) {
  auto self_ = unpack(self, "self", 0);
  auto _any_requires_grad = compute_requires_grad( self );

  (void)_any_requires_grad;
  std::shared_ptr<ForeachAcosBackward0> grad_fn;
  if (_any_requires_grad) {
    grad_fn = std::shared_ptr<ForeachAcosBackward0>(new ForeachAcosBackward0(), deleteNode);
    grad_fn->set_next_edges(collect_next_edges( self ));
    grad_fn->self_ = make_saved_variable_list(self);
    grad_fn->self_size_ = self.size();
  }
...
  if (grad_fn) {
      set_history(flatten_tensor_args( result ), grad_fn);
  }
  return result;
}
```

As I wrote above, `result[i]` only needs `self[i]` so something which could fix the behavior would be follows:

```cpp
// torch/csrc/autograd/generated/Functions.h
struct TORCH_API ForeachAcosBackward0 : public TraceableFunction {
  using TraceableFunction::TraceableFunction;
  variable_list apply(variable_list&& grads) override;
  std::string name() const override { return "ForeachAcosBackward0"; }
  void release_variables() override {
    std::lock_guard<std::mutex> lock(mutex_);
    self_.reset_data();
  }
  SavedVariable self_;
};

// torch/csrc/autograd/generated/Functions.cpp
variable_list ForeachAcosBackward0::apply(variable_list&& grads) {
  std::lock_guard<std::mutex> lock(mutex_);
  TORCH_CHECK(!self_released_, ERR_BACKWARD_TWICE);
  IndexRangeGenerator gen;
  auto self_ix = gen.range(1);
  variable_list grad_inputs(gen.size());
  auto self = self_.unpack();
  if (task_should_compute_output({ self_ix })) {
    auto grad_result = grads * -((-self * self + 1).rsqrt()).conj());
    copy_range(grad_inputs, self_ix, grad_result);
  }
  return grad_inputs;
}
```

```cpp
// torch/csrc/autograd/generated/VariableTypeEverything.cpp
::std::vector<at::Tensor> _foreach_acos(c10::DispatchKeySet ks, at::TensorList self) {
  auto self_ = unpack(self, "self", 0);
  auto _any_requires_grad = compute_requires_grad( self );

  (void)_any_requires_grad;
  std::shared_ptr<ForeachAcosBackward0> grad_fn;
  std::vector<std::shared_ptr<ForeachAcosBackward0>> grad_fns;
  grad_fns.reserve(self.size());
  if (_any_requires_grad) {
    for (const auto & i : c10::irange(self.size()) {
      grad_fn = std::shared_ptr<ForeachAcosBackward0>(new ForeachAcosBackward0(), deleteNode);
      grad_fn->SavedVariable(self[i], false);
      grad_fns.emplace_back(grad_fn);
    }
  }
...
  if (grad_fn) {
      for (const auto & i : c10::irange(result.size()) {
          set_history(result[i], grad_fn[i]);
      }
  }
  return result;
}
```

**Comment by crcrpar** [comment]
I noticed that codegen'ed `ForeachXXXBackwardX<overload>` don't have `copy_properties` in `python_functions`...

**Comment by ezyang** [comment]
This doesn't seem that surprising to me, since normally a multi-output function generates a single backward function that handles all of the outputs.

**Comment by soulitzer** [comment]
The snippet in the issue description is expected yes, but this seems like a tricky problem, and I think the proposed solution is a interesting approach toward solving it.

I think it's a tricky problem because the desired semantics of foreach is wrt autograd are unclear to me. Here are a few ideas, curious what others think:
1)  respect the requires_grad-ness of the inputs. The state today is that as long as a single input requires_grad, all outputs will require grad, we may also save extraneous inputs/outputs. This constraint proposes that only the tensors corresponding to the inputs that actually require grad should be saved, and only the outputs that correspond to the inputs that require grad should require grad.
2) allow backward wrt particular inputs of foreach only
3)  provide enhanced performance during backward as well by doing another foreach during backward (e.g. _foreach_sin -> _foreach_cos)

The state today is that we have a single node generated. The proposed solution above says that we want multiple nodes. 
- Single node (basically what we have today): (1) is harder (2) is not feasible, (3) can be done easily
- Multi node (as seen in the proposed solution): (1) and (2) are easier, (3) is not feasible

The trade off is essentially between (2) and (3) which seem mutually exclusive. I'm leaning toward supporting (2) because it differentiates foreach from nested tensor a little more.

**Comment by crcrpar** [comment]
Thank you both of you for comments.
I'm yet to clearly understand how the codegen is working so I do appreciate comments and feedbacks.

> the desired semantics of foreach is wrt autograd are unclear to me

from my perspective it should be similar to native functions given that foreach functions call a reference native function iteratively for certain inputs (by a reference native function I mean e.g. `torch.add` for `torch._foreach_add`).

**Comment by ezyang** [comment]
I mean, if we wanted to break up the autograd graph for each of the outputs, we COULD do that. There's not very much precedent for doing this, since typically you get a better fused kernel if you handle all of the outputs at once, but it is very much possible. It does seem like you could potentially get a better fused kernel for foreach if you handle it in one go, if all of the inputs require derivatives, but I guess it's more likely the requires grad will be sparse and you're not going to get much foreach benefit (or maybe I'm wrong?)

**Comment by soulitzer** [comment]
@crcrpar If you mean to say that autograd in the case of `torch._foreach_add` should behave AS IF we did a for loop of `torch.add`, that would make sense to me. Ideally the foreach optimization is transparent to user in this respect.

Given this desired semantic, our two options are: (A) generate multiple nodes, OR (B) generate a single node, but have this node emulate the behavior of multiple nodes.

(A) seems more straightforward for sure, but I think its worth investigating (B) some more since we'd ideally like to fuse the kernel for backward as well as Ed suggests (I'm not sure we care about fusing compilers too much in the foreach case, but that said, situation seems worse in eager anyway.)

If we were to break down: "autograd should behave as if we did a loop of `torch.add`", the requirements are mainly (1) and (2) as in my comment above. (Obviously we cannot expect the behavior to be entirely the same, like the user is capable of introspecting how many Nodes there are generated)

I think satisfying constraint (1) isn't too bad. You just need some codegened logic conditionally saving and setting requires grad. (2) I said previously that it wasn't feasible but I'm not actually sure, it may require modifications to the engine (still thinking about this).

**Comment by soulitzer** [comment]
I thought about how (B) can satisfy (2) some more, and it seems to require extensive changes to the engine indeed. If a single node is supposed to emulate the topology of multiple nodes, it fundamentally changes how we traverse through the graph. Now as we traverse through the graph, we not only need to know the node we are currently at, but also additional state to track the incoming edge we originally reached the node through. I feel like we want to avoid this for the moment.

If we now consider (A) and (B) when (2) isn't satisfied they have similar complexity. It's just a matter of whether we care more about doing foreach backward in a fused way (eagerly) vs. allowing someone to backward flexibly with inputs= without incurring unnecessary computation. 

(B) seems to be a better default. No matter what your use case is, you'll always run into the issue of having to backward through that foreach node again. But not everyone runs .grad() or .backward() with inputs=. For us to prioritize the latter case, we'd have to believe that (1) a good amount of users will backward with inputs= AND (2) that perf-wise the extra computation unnecessarily done exceeds the savings achieved by fusing the foreach backward.

Maybe more investigation is worth it to make a better decision, but at the moment, a possible plan is: 
1) Improve (B) to satisfy constraint (1), i.e., update the single node case to respect the requires_grad of the inputs
2) Update its foreach's backward to be another foreach

If we wanted to do more investigation, that would involve looking at the use cases a differentiable foreach is is targeting (likely MAML use cases where we are differentiating through the optimizer pass).

**Comment by ngimel** [comment]
>but I guess it's more likely the requires grad will be sparse

Why would that be the case? I'd expect list inputs to have similar requires_grad status. Of course most actual usages come from optimizers where it doesn't matter, but suppose we lifted contiguous storage requirement on NestedTensors and backed them with for_each implementations - I would expect NestedTensor elements to have common requires_grad, no? In fact I think NestedTensor only allows a single requires_grad for that whole batch.

**Comment by soulitzer** [comment]
From discussion with with @albanD, we should follow a mixed approach:
- for out-of-place op we're okay with the trade off of doing single node, i.e. follow proposal per https://github.com/pytorch/pytorch/issues/95431#issuecomment-1451098438 
- for in-place and view ops we should be doing multiple nodes since handling in-place over view may be infeasible in the single-node case

**Comment by crcrpar** [comment]
I though we can close this given that I understand that we're okay with out-of-place foreach functions have entire tensorlists attached to a grad_fn and in-place foreach functions have grad_fn per pair of tensors of the same index.

Please correct me if I'm wrong :)
